### PR TITLE
[dd] enable change ddse from innodb to rocksdb

### PR DESCRIPTION
**Pull request by luqun**
Summary:

The change is to "upgrade" DD tables from innodb SE to rocksdb SE.

Steps for a DD table involving DDSE change: apply to all DD tables 
```
- create target DD tables into <upgrade> schema by execute "Create table"
- 'logical copy': execute "INSERT INTO <upgrade>.<Table> SELECT * FROM mysql.<Table>" 
- "rename" mysql.<Table> to <temporary>.<Table> and "rename" <upgrade>.<Table> to mysql.<Table>
```

Workflow for innodb SE to rocksdb SE changes
```
- create a schema for target DD tables(such as <upgrade>) and create another schema for temporary usage(such as <temporary>
- upgrade dd_properties table
- upgrade other DD tables
- update DD_ENGINE/VERSION property in dd_properties table and  commit changes with one trx
- drop <temporary> schema
```

The above workflow for INNODB->ROCKSDB DDSE change is finished with two writebatch commit first and one transaction commit
- Two writebatch: see update_myrocks_table_names()
   - The first write batch is to fix/update/rename myrocks internal DD for dd_property table(rename <upgrade>.dd_properties to mysql.dd_properties); after commit, table_exists_in_engine for ROCKSDB will return true; while before commit, table_exists_in_engine for ROCKSDB will return false
    - the second batch is also to fix/update/rename myrocks internal DD for other DD tables.
- One transaction:  it hold create tables, "rename" table, "logical copy" and committed in  update_versions()

 Whether to do DDSE change is controlled by 3 DDSE values:
- default_dd_storage_engine: It is a mysql variable and its value is defined in CNF file or default value if not specified. Use V1 for reference
- DD_ENGINE property in mysql.properties table: its value is stored during last step of  previous DDSE change or during boostrap, Use V2 for reference
- table_exists_in_engine handle API: runtime only value. before open first table(dd_properties), guess that if one SE hold dd_properties table, then this SE might be DDSE. Use V3 for reference

If V1, V2, V3 are equal, then SKIP DD table change; otherwise, try DD table change

During DDSE change, if crash occurs, what's next?

Crash Scenarios during INNODB->ROCKSDB DDSE change
- Crash before first writebatch commit:  after crash, during restart, its V1 is ROCKSDB, V2 is INNODB, V3 is INNODB, upgrade DD tables again.
- Crash before second writebatch commit or  before Transaction commit: after crash, during restart, its V1 value is ROCKSDB, V2 value is INNODB, V3 value is ROCKSDB, upgrade DD tables again..
- crash after Transaction commit: after crash, during restart, its V1 value is ROCKSDB, V2 value is ROCKSDB and V3 is ROCKSDB, SKIP upgrade DD tables

Remaining issues:
- during ddse change, there a crash. after restart, CNF changed.

### CODE REVIEW

File: sql/dd/impl/upgrade/dd.cc [hunk 1]
```diff
@@ -1031,6 +1032,84 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+                                 Object_id target_table_schema_id,
+                                 const String_type &target_table_schema_name,
+                                 Object_id actual_table_schema_id) {
+  // only upgrade dd_properties table for dd_engine table
+  // during dd_version change, dd_properties table keep same
+  if (!bootstrap::DD_bootstrap_ctx::instance().is_dd_engine_change())
+    return false;
+
+  // create <upgrade>.dd_properties table with default_ddse
+  dd::tables::DD_properties::instance().set_actual_engine(get_dd_engine_name());
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: The code uses mixture of `auto` and explicit types, i.e. handlerton type variables tend to be explicit and handler ones auto. Please consider using a single style, preferably `auto`

File: sql/dd/impl/upgrade/dd.cc [hunk 2]
```diff
@@ -1031,6 +1032,84 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+
+  // myrocksdb has its own data dictionary, call handle rename API to
+  // update myrocksdb own data dictionary
+  if (default_dd_storage_engine == DEFAULT_DD_ROCKSDB) {
+    handlerton *rocksdb_ddse = ha_resolve_by_legacy_type(thd, DB_TYPE_ROCKSDB);
+    if (!rocksdb_ddse) {
+      assert(false);
+      return true;
+    }
+    auto handler = get_new_handler(nullptr, false, thd->mem_root, rocksdb_ddse);
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: Even with `auto` it's a good idea to show pointers and references:
```suggestion
    auto *handler = get_new_handler(nullptr, false, thd->mem_root, rocksdb_ddse);
```

File: sql/dd/impl/upgrade/dd.cc [hunk 3]
```diff
@@ -1169,6 +1256,38 @@ bool upgrade_tables(THD *thd) {
...
+    if (!handler) {
+      assert(false);
+      return true;
+    }
+
+    for (std::set<String_type>::const_iterator it = create_set.begin();
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
    for (auto it = create_set.cbegin();
```

File: sql/dd/impl/upgrade/dd.cc [hunk 4]
```diff
@@ -1169,6 +1256,38 @@ bool upgrade_tables(THD *thd) {
...
+         it != create_set.end(); ++it) {
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
         it != create_set.cend(); ++it) {
```

**Review by laurynas-biveinis — COMMENTED** [review]
One complication with dd_properties is that on MyRocks -> InnoDB DDSE change InnoDB will have an internal table `dd_properties_placeholder`, which is inert but importantly reserves the hardcoded InnoDB IDs for the `dd_properties` table. This table will have to be dropped and the `dd_properties` in InnoDB (in its temp table table) will have to be ensured to take those IDs

**Comment by laurynas-biveinis** [comment]
@luqun , is this ready for review?

**Comment by luqun** [comment]
> @luqun , is this ready for review?
Yes. maybe I shouldn't force push?

**Comment by laurynas-biveinis** [comment]
> @luqun , is this ready for review?
> Yes. maybe I shouldn't force push?

Force pushes are fine with me

File: sql/dd/impl/upgrade/dd.cc [hunk 5]
```diff
@@ -1031,6 +1032,90 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  upgrade dd_properties table during dd_engine change
+
+  @param  thd                      Thread context.
+  @param  mysql_schema_id          Id of the 'mysql' schema.
+  @param  target_table_schema_id   Id of the schema where the tables in the
+                                   create_set are located.
+  @param  target_table_schema_name Name of the schema where the tables in the
+                                   create_set are located.
+  @param  actual_table_schema_id   Id of the schema where the tables in the
+                                   remove_set will be moved.
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: remove_set?

File: sql/dd/impl/upgrade/dd.cc [hunk 6]
```diff
@@ -1031,6 +1032,90 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  std::set<String_type> dd_properties_set({dd_property_table_name});
+  if (update_object_ids(thd, dd_properties_set, dd_properties_set,
+                        mysql_schema_id, target_table_schema_id,
+                        target_table_schema_name, actual_table_schema_id)) {
+    assert(false);
+    return true;
+  }
+  dd::end_transaction(thd, false);
+
+  // myrocksdb has its own data dictionary which contain table name, call
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
  // myrocksdb has its own data dictionary which contains table name, call
```

File: sql/dd/impl/upgrade/dd.cc [hunk 7]
```diff
@@ -1031,6 +1032,90 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  // handler rename API to fix these table name
+  if (default_dd_storage_engine == DEFAULT_DD_ROCKSDB) {
+    handlerton *rocksdb_ddse = ha_resolve_by_legacy_type(thd, DB_TYPE_ROCKSDB);
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
    auto *rocksdb_ddse = ha_resolve_by_legacy_type(thd, DB_TYPE_ROCKSDB);
```

File: sql/dd/impl/upgrade/dd.cc [hunk 8]
```diff
@@ -1031,6 +1032,90 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+                         target_table_schema_name.c_str(),
+                         dd_property_table_name.c_str(), "", 0);
+    // new table name under mysql schema
+    build_table_filename(lc_to, sizeof(lc_to) - 1, MYSQL_SCHEMA_NAME.str,
+                         dd_property_table_name.c_str(), "", 0);
+    handler->ha_rename_table(lc_from, lc_to, nullptr, nullptr);
+    destroy(handler);
+  }
+
+  // dd_properties has been upgrade to target engine
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
  // dd_properties has been upgraded to target engine
```

File: sql/dd/impl/upgrade/dd.cc [hunk 9]
```diff
@@ -1169,6 +1262,38 @@ bool upgrade_tables(THD *thd) {
...
+      build_table_filename(lc_from, sizeof(lc_from) - 1,
+                           target_table_schema_name.c_str(), (*it).c_str(), "",
+                           0);
+      build_table_filename(lc_to, sizeof(lc_to) - 1, MYSQL_SCHEMA_NAME.str,
+                           (*it).c_str(), "", 0);
+      handler->ha_rename_table(lc_from, lc_to, nullptr, nullptr);
+    }
+    destroy(handler);
+
+    if (dd::execute_query(thd, "set global rocksdb_force_flush_memtable_now=1"))
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: Why is this needed?

File: sql/dd/impl/upgrade/dd.cc [hunk 10]
```diff
@@ -1031,6 +1032,90 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+                        mysql_schema_id, target_table_schema_id,
+                        target_table_schema_name, actual_table_schema_id)) {
+    assert(false);
+    return true;
+  }
+  dd::end_transaction(thd, false);
+
+  // myrocksdb has its own data dictionary which contain table name, call
+  // handler rename API to fix these table name
+  if (default_dd_storage_engine == DEFAULT_DD_ROCKSDB) {
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: Duplicated but not exactly (single table vs set) with `upgrade_tables` below. Please see if something meaningful could be extracted as a helper, for example lines 1101 (`char lc_from`) to 1110 (`handler->ha_rename_table`) block.

**Review by laurynas-biveinis — COMMENTED** [review]
There is a RocksDB submodule change in the PR, most likely non-intentional?

File: sql/dd/impl/upgrade/dd.cc (hunk 9 above)
Comment (luqun) [reply sql/dd/impl/upgrade/dd.cc]: For correctly, it isn't needed..it is easier to diagnostic if something goes wrong(check SST file is easier to walk through LSM). let me remove it first and may add it back in another diff.

File: sql/dd/impl/upgrade/dd.cc (hunk 10 above)
Comment (luqun) [reply sql/dd/impl/upgrade/dd.cc]: add another helper function

File: sql/dd/impl/upgrade/dd.cc [hunk 11]
```diff
@@ -1031,6 +1032,124 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  Update myrocks internal data during dd_engine change from innodb to rocksdb
+
+  @param  thd                      Thread context.
+  @param  old_schema_name          Name of the old(from) schema
+  @param  new_schema_name          Name of the new(to) schema
+  @param  table_names              table name set to be updated
+
+  @returns false if success, true otherwise.
+*/
+bool update_myrocks_table_name(THD *thd, const String_type &old_schema_name,
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]:
```suggestion
bool update_myrocks_table_names(THD *thd, const String_type &old_schema_name,
```
because it's multiple names in a set

File: sql/dd/impl/upgrade/dd.cc [hunk 12]
```diff
@@ -1031,6 +1032,124 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  if (!rocksdb_ddse) {
+    assert(false);
+    return true;
+  }
+  auto *handler = get_new_handler(nullptr, false, thd->mem_root, rocksdb_ddse);
+  if (!handler) {
+    assert(false);
+    return true;
+  }
+  int err = false;
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: Don't combine `int` and `false`, please change to either `bool` or `0`.

File: sql/dd/impl/upgrade/dd.cc [hunk 13]
```diff
@@ -1031,6 +1032,124 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  for (auto table_name : table_names) {
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: The lack of reference creates a new value every time
```suggestion
  for (const auto &table_name : table_names) {
```

File: sql/dd/impl/upgrade/dd.cc [hunk 14]
```diff
@@ -1031,6 +1032,124 @@ bool update_object_ids(THD *thd, const std::set<String_type> &create_set,
...
+  std::set<String_type> dd_properties_set({dd_property_table_name});
+  if (update_object_ids(thd, dd_properties_set, dd_properties_set,
+                        mysql_schema_id, target_table_schema_id,
+                        target_table_schema_name, actual_table_schema_id)) {
+    assert(false);
+    return true;
+  }
+  dd::end_transaction(thd, false);
+
+  // myrocks has its own data dictionary which contains table name, call
```
Comment (laurynas-biveinis) [inline sql/dd/impl/upgrade/dd.cc]: Redundant comment now that `update_myrocks_table_name` exists and has its own table header comment. Maybe add a bit more info there if needed

**Review by laurynas-biveinis — COMMENTED** [review]
LGTM! 

The "during ddse change, there a crash. after restart, CNF changed." part can be addressed in a follow-up PR, if you prefer.

The PR description now contains useful information. Could you please copy it to Wiki?

**Comment by luqun** [comment]
committed
